src/libsync/creds/oauth.cpp

```diff
@@ -182,25 +171,39 @@ void OAuth::startAuthentification()
 
 void OAuth::refreshAuthentification(const QString &refreshToken)
 {
-    QObject::connect(this, &OAuth::fetchWellKnownFinished, this, [this, &refreshToken]() {
+    connect(this, &OAuth::fetchWellKnownFinished, this, [this, refreshToken] {
         auto job = postTokenRequest({ { QStringLiteral("grant_type"), QStringLiteral("refresh_token") },
             { QStringLiteral("refresh_token"), refreshToken } });
-        QObject::connect(job, &SimpleNetworkJob::finishedSignal, this, [this](QNetworkReply *reply) {
-            auto jsonData = reply->readAll();
+        connect(job, &SimpleNetworkJob::finishedSignal, this, [this, refreshToken](QNetworkReply *reply) {
+            const auto jsonData = reply->readAll();
+            QString accessToken;
+            QString newRefreshToken = refreshToken;
             QJsonParseError jsonParseError;
+            // https://developer.okta.com/docs/reference/api/oidc/#response-properties-2
             const QJsonObject json = QJsonDocument::fromJson(jsonData, &jsonParseError).object();
-            const QString accessToken = json["access_token"].toString();
-            if (jsonParseError.error != QJsonParseError::NoError || json.isEmpty()) {
-                // Invalid or empty JSON: Network error maybe?
-                qCWarning(lcOauth) << "Error while refreshing the token" << reply->errorString() << jsonData << jsonParseError.errorString();
-            } else if (accessToken.isEmpty()) {
-                // If the json was valid, but the reply did not contain an access token, the token
-                // is considered expired. (Usually the HTTP reply code is 400)
-                qCDebug(lcOauth) << "Expired refresh token. Logging out";
-                Q_EMIT refreshFinished(QString(), QString());
+            const QString errorReason = errorFromJson(json);
+            if (!errorReason.isEmpty()) {
+                if (errorReason == QStringLiteral("invalid_grant")) {
```

Comment:
Can I assume this response or should I rather check for a http error code?

Reply:
To answer my question no.
Is there a proper way to detect a rejection getting proper json without a token feels wrong?